pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +229,318 @@ def compute_jacobian(self, tapes: Batch):
+    ``qml.devices.Device.compute_derivatives``.
+
+    Args:
+
+        device (Union[pennylane.Device, pennylane.devices.Device]): the device for execution and derivatives.
+            Must support first order gradients with the requested configuration.
+        execution_config (pennylane.devices.ExecutionConfig): a datastructure containing the options needed to fully
+           describe the execution. Only used with :class:`pennylane.devices.Device` from the new device interface.
+        gradient_kwargs (dict): a dictionary of keyword arguments for the gradients. Only used with a :class:`~.pennylane.Device`
+            from the old device interface.
+
+    **Examples:**
+
+    >>> device = qml.device('default.qubit')
+    >>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
+    >>> jpc = DeviceDerivatives(device, config, {})
+
+    This same class can also be used with the old device interface.
+
+    >>> device = qml.device('lightning.qubit', wires=5)
+    >>> gradient_kwargs = {"method": "adjoint_jacobian"}
+    >>> jpc_lightning = DeviceDerivatives(device, gradient_kwargs=gradient_kwargs)
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and Jacobians for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+    In the current execution pipeline, only one batch will be used per instance, but a size of 10 adds some extra
+    flexibility for future uses.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents. So batches
+    with identically looking :class:`~.QuantumScript` s that are different instances will be cached separately.
+
+    When a forward pass with :meth:`~.execute` is called, both the results and the jacobian for the object are stored.
+
+    >>> tape = qml.tape.QuantumScript([qml.RX(1.0, wires=0)], [qml.expval(qml.PauliZ(0))])
+    >>> batch = (tape, )
+    >>> with device.tracker:
+    ...     results = jpc.execute(batch )
+    >>> results
+    (0.5403023058681398,)
+    >>> device.tracker.totals
+    {'execute_and_derivative_batches': 1, 'executions': 1, 'derivatives': 1}
+    >>> jpc._jacs_cache
+    LRUCache({5660934048: (array(-0.84147098),)}, maxsize=10, currsize=1)
+
+    Then when the vjp, jvp, or jacobian is requested, that cached value is used instead of requesting from
+    the device again.
+
+    >>> with device.tracker:
+    ...     vjp = jpc.compute_vjp(batch , (0.5, ) )
+    >>> vjp
+    (array([-0.42073549]),)
+    >>> device.tracker.totals
+    {}
+
+    **Lack of shot vector support:**
+
+    Given we currently do not have any examples of device derivatives with shots vectors (or even finite shots),
+    we are bypassing support for partitioned shots.
```

Comment:
@albi3ro what do you mean by this?

> (or even finite shots)

Reply:
Is this because the adjoint method for lightning only supports `shots=None`?

Reply:
adjoint on both default qubit and lightning are only supported for `shots=None`.

Reply:
I wonder if it could be tested with minimal effort through mocking? E.g., create a mock device that does nothing but return an arbitrary ndarray for finite shot jacobian for any circuit.

Or, mock default.qubit to return an arbitrary ndarray if the finite shot device jacobian is requested.

Reply:
E.g., don't test that the result is correct, just check that the pipeline works.

Reply:
I've added a `ParamShiftDerivativesDevice` to `tests/param_shift_device_dev.py`.  This device uses parameter shift to product "device" derivatives. With this, we can test device derivatives and jacobian products with shot vectors.

I was able to add it for the vjp, since I had some reference to work with in `autograd.py`, but I'm still trying to figure out how to integrate jvp's with shot vectors.

Reply:
> I'm still trying to figure out how to integrate jvp's with shot vectors

@albi3ro what was the resolution here?